File: GetMSEBetaNorm/ExtractMain.m
%ExtractMain
%功能：计算跨中单元各种损伤状况下，各单元的模态应变能及模态应变能变化率（暂时只算其中特定阶数模态）

%mse:模态单元应变能
tic;
clear;clc;

%损伤单元弹性模量的损伤系数服从N(MU,SIGMA)
MU=0.9; 
SIGMA=MU*0.01;

xlsFileName = 'mseResult.xlsx';

firstFileName='firstFile.inp';      %第1个拼接文件
secondFileName='secondFile.inp';
thirdFileName='thirdFile.inp';

sourcePath='d:\sen_result.txt'; %ansys计算结果存储文件路径
ansysPath='C:\progra~1\ANSYSI~1\v160\ansys\bin\winx64\ansys160.exe';    %ansys安装路径

inputFile='BeamExampleNoDamageSetPositionNoDama.inp';   %ansys cmd 中输入文件
outputFile='d:\ansysOutput.txt';    %ansys cmd中输出文件

%----------------------------------------------
%计算无损情况下各个单元的模态应变能
cmdStr=[ansysPath ' -b -p ane3fl -i ' inputFile ' -o ' outputFile];
system(cmdStr);
mseData=load(sourcePath);     %读取数据
mseNoDama=mseData;    %无损工况各单元单元应变能

%----------------------------------------------

inputFile='BeamExampleDamageCombine.inp';   %ansys cmd 中输入文件

loopCount=1000;   %模拟次数
mseDama=[];  %损伤后各单元模态应变能
i=1;
while i<=loopCount
    R = normrnd(MU,SIGMA);    %损伤系数,0.9左右
    %防止随机数中出现R<0的情况
    while R<0
        R = normrnd(MU,SIGMA);
    end
    midFileStr=['dF=' num2str(R)];     %第2个文件的内容
    
    fid=fopen(secondFileName,'w');
    fprintf(fid,'%s\n',midFileStr);
    fclose(fid);
    
    fid1=fopen(firstFileName,'r');
    fid2=fopen(secondFileName,'r');
    fid3=fopen(thirdFileName,'r');
    Data1=fread(fid1);
    Data2=fread(fid2);
    Data3=fread(fid3);
    
    fid=fopen(inputFile,'w');   %合并写入的文件
    
    fwrite(fid,Data1);
    fwrite(fid,Data2);
    fwrite(fid,Data3);
    fclose(fid1);
    fclose(fid2);
    fclose(fid3);
    
    fclose(fid);
    
    cmdStr=[ansysPath ' -b -p ane3fl -i ' inputFile ' -o ' outputFile];
    
    system(cmdStr);
    
    mseData=load(sourcePath);     %读取数据
    
    mseDama=[mseDama mseData];
    
    disp(['progress:' num2str(i) 'th:' num2str(i*100/loopCount) '%']);
    toc;
    i=i+1;
end

nElems=size(mseDama,1);     %单元个数



sheetHeader={'id','seneNoDama'};
for i=1:loopCount
    sheetHeader{i+2}=i;
end

lst(1:nElems,1)=1:nElems;   %列表


sheetName=num2str(nElems);

xlRange = 'A1';
xlswrite(xlsFileName,sheetHeader,sheetName,xlRange)
xlRange = 'A2';
xlswrite(xlsFileName,lst,sheetName,xlRange)
xlRange = 'B2';
xlswrite(xlsFileName,mseNoDama,sheetName,xlRange)
xlRange = 'C2';
xlswrite(xlsFileName,mseDama,sheetName,xlRange)

%下一张表
%-------------------
sheetHeader={'id'};
for i=1:loopCount
    sheetHeader{i+1}=i;
end

div=repmat(mseNoDama,1,loopCount);  %除数扩展
betaResult=(mseDama-div)./div;
sheetNo = 2;    %存储"各单元各损伤工况模态应变能变化率"

sheetName=[num2str(nElems) '_beta'];

xlRange = 'A1';
xlswrite(xlsFileName,sheetHeader,sheetName,xlRange)
xlRange = 'A2';
xlswrite(xlsFileName,lst,sheetName,xlRange)
xlRange = 'B2';
xlswrite(xlsFileName,betaResult,sheetName,xlRange)

toc;

File: guanhe/GuanheAnsysFilesZRM/ExtractMain.py
# -*- coding: utf-8 -*-
"""
Created on Sun Sep 24 18:42:38 2017

@author: bdl
"""

import numpy as np
import os


#数组使用参考：
#http://blog.csdn.net/taoyanqi8932/article/details/52703686

loopCount=2    #循环模拟次数

nElems=178    #输出单元
controlOption=np.zeros((nElems)) 
#controlOption=np.array([0,0,1])  #随机控制选项
                        #举例：[0 0 1]表示第1、2个单元不进行随机控制，第3个单元进行随机控制
controlOption[88]=1

#controlOption==0对应项不予考虑
MU=np.ones((nElems))
MU[88]=0.90
#MU=np.array([1,1,0.9])
SIGMA=MU*0.01

xlsFileName = 'mseResult.xlsx'

firstFileName='firstFile.inp'      #第1个拼接文件
secondFileName='secondFile.inp'
thirdFileName='thirdFile.inp'

sourcePath='e:\\sen_result.txt' #ansys计算结果存储文件路径
ansysPath='C:\\progra~1\\ANSYSI~1\\v160\\ansys\\bin\\winx64\\ansys160.exe'    #ansys安装路径

inputFile='GuanheNoDama.inp'   #ansys cmd 中输入文件（无损结构）
outputFile='e:\\ansysOutput.txt'    #ansys cmd中输出文件

#----------------------------------------------
#计算无损情况下各个单元的模态应变能
cmdStr=ansysPath + ' -b -p ane3fl -i ' + inputFile + ' -o ' + outputFile

os.system(cmdStr)

mseData = np.loadtxt(sourcePath)    #读取数据  
mseNoDama=np.transpose([mseData])    #无损工况各单元单元应变能 
#----------------------------------------------

inputFile='GuanheDamageCombine.inp'   #ansys cmd 中输入文件（组合文件）

#mseDama=[]  损伤后各单元模态应变能
midFileStr=''  #中间文件字符串
i=1

while i<=loopCount:
    R=np.multiply(SIGMA,np.random.randn(1, MU.shape[0])) + MU    #损伤系数;MU.shape[0]:MU列数
      
    #防止随机数中出现R<0的情况
    #必须每个都>=0
    #while R<0:
    #    R=multiply(SIGMA,np.random.randn(1, MU.shape[0])) + MU    #损伤系数;MU.shape[0]:MU列数
    
    for j in range(0,len(R[0])):
        if controlOption[j]==0:
             R[0,j]=1

    midFileStr=[]
    
    midFileStr=''
    for j in range(0,len(R[0])):
        midFileStr=midFileStr+'dFactor('+str(j+1)+')='+str(R[0,j])+chr(13)+chr(10)
                
    
    with open(secondFileName, 'w') as f:
        f.write(midFileStr);
    
    with open(firstFileName,'rb') as f1:
        data1 = f1.read()
    
    with open(secondFileName,'rb') as f2:
        data2 = f2.read()
    
    with open(thirdFileName,'rb') as f3:
        data3 = f3.read()
    
    with open(inputFile, 'wb') as f:
        f.write(data1)
        f.write(data2)
        f.write(data3)
    
    cmdStr=ansysPath + ' -b -p ane3fl -i ' + inputFile + ' -o ' + outputFile
    
    os.system(cmdStr)
    
    mseData = np.loadtxt(sourcePath) 
    mseData=np.transpose([mseData])
    
    if i==1:
        mseDama=mseData    #损伤后各单元模态应变能
    else:
        mseDama=np.hstack((mseDama,mseData))
        
  
    print('progress:' + str(i) + 'th:' + str(i*100/loopCount)+'%')
    
    i=i+1

import pandas as pd

nElems=mseDama.shape[0]      #单元个数

sheetHeader=['seneNoDama']   #第1个表头id需要自己补

for i in range(0,loopCount):
    sheetHeader.append(i+1);

lst=[n for n in range(1,nElems+1)]   #列表

sheetName=str(nElems)

mseCombine=np.hstack((mseNoDama,mseDama))

# prepare for data
data_df = pd.DataFrame(mseCombine)
# change the index and column name
data_df.columns = sheetHeader
data_df.index = lst

# create and writer pd.DataFrame to excel
writer = pd.ExcelWriter(xlsFileName)
data_df.to_excel(writer,sheetName,float_format='%.5f') # float_format 控制精度
#writer.save() 

#下一张表
#-------------------
sheetHeader=[]    #空list,#第1个表头id需要自己补
for i in range(0,loopCount):
    sheetHeader.append(i+1)

div=mseNoDama#除数扩展,对应matlabdiv=repmat(mseNoDama,1,loopCount)  
betaResult=(mseDama-div)/div    #对应matlab ./

#第2张表存储"各单元各损伤工况模态应变能变化率"


sheetName=str(nElems)+'_beta'

# prepare for data
data_df2 = pd.DataFrame(betaResult)
# change the index and column name
data_df2.columns = sheetHeader
data_df2.index = lst

# create and writer pd.DataFrame to excel
#writer = pd.ExcelWriter(xlsFileName)
data_df2.to_excel(writer,sheetName,float_format='%.5f') # float_format 控制精度
writer.save() 

"""
xlsxFileName='fitResult.xlsx'
sheetName='Sheet1'

data=xlrd.open_workbook(xlsxFileName)
sh=data.sheet_by_name(sheetName)

print(sh.cell_value(0,0))
"""